lib/ramble/ramble/keywords.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2022-2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+from enum import Enum
+
+import llnl.util.tty as tty
+
+import ramble.error
+
+
+class Keywords(object):
+    """Class to represent ramble known keywords.
+
+    Each keyword can have a type. Valid types are in the class level
+    'key_types' variable as an enum.
+
+    A map of keys, and their types is provided in the class level 'keys'
+    variable. This enforces that each keyword can only have a single type.
+
+    Current key types are:
+    - Reserved: Ramble defines these, and a user should not be allowed to define them
+    - Optional: Ramble can function with a definition from the user but it isn't required
+    - Required: Ramble requires a definition for these. Ramble will try to set sensible defaults,
+                but it might not be possible always.
+
+    """
+
+    key_types = Enum('types', ['reserved', 'optional', 'required'])
+
+    keys = {
+        'application_name': key_types.reserved,
+        'application_run_dir': key_types.reserved,
+        'application_input_dir': key_types.reserved,
+        'workload_name': key_types.reserved,
+        'workload_run_dir': key_types.reserved,
+        'workload_input_dir': key_types.reserved,
+        'experiment_name': key_types.reserved,
+        'experiment_run_dir': key_types.reserved,
+        'log_dir': key_types.reserved,
+        'log_file': key_types.reserved,
+        'err_file': key_types.reserved,
+        'command': key_types.reserved,
+        'spack_env': key_types.reserved,
+
+        'spec_name': key_types.optional,
+
+        'n_ranks': key_types.required,
+        'n_nodes': key_types.required,
+        'processes_per_node': key_types.required,
+        'n_threads': key_types.required,
+        'batch_submit': key_types.required,
+        'mpi_command': key_types.required,
+    }
+
+    def __init__(self):
+        # Define class attributes for all of the keys
+        for key in self.keys.keys():
+            setattr(self, key, key)
```

Comment:
Can you remind me what the motivation was to do it this way instead of base classing enum? (ie `class Keywords(Enum):`. I'm fine either way, I was just surprised to see it this way round, but I like that we avoid the `.name`

Reply:
That was really just it. If I used enum, the default is to give the ID, and I didn't love having to append `.name` to all of the key names just to get the name.

The enum didn't really buy me anything beyond having a standard class for them, but I needed to wrap the logic anyway so I decided it would just confuse the implementation more.

Reply:
Also, it doesn't seem like there is a nice way to union enums, and with a single enum it's kind of gross to manage the types associated with each key.